### PR TITLE
Add ability to exclude Blocks in AlignedLattice

### PR DESCRIPTION
**Pull request by mar-celine**
## Proposed changes

Allows one to create a cubical shell domain (see factory test for possible input file parameters)

### Types of changes:

- [ ] Bugfix
- [ ] New feature

### Component:

- [ ] Code
- [ ] Documentation
- [ ] Build system
- [ ] Continuous integration

### Code review checklist

- [ ] The PR passes all checks, including unit tests, `clang-tidy` and `IWYU`.
  For instructions on how to perform the CI checks locally refer to the [Dev
  guide on the Travis CI](https://spectre-code.org/travis_guide.html).
- [ ] The code is documented and the documentation renders correctly. Run
  `make doc` to generate the documentation locally into `BUILD_DIR/docs/html`.
  Then open `index.html`.
- [ ] The code follows the stylistic and code quality guidelines listed in the
  [code review guide](https://spectre-code.org/code_review_guide.html).

### CODE REVIEW

File: src/Domain/Creators/AlignedLattice.hpp [hunk 1]
```diff
@@ -59,18 +59,34 @@ class AlignedLattice : public DomainCreator<VolumeDim, TargetFrame> {
...
-      typename IsPeriodicIn::type is_periodic_in,
-      typename InitialRefinement::type initial_refinement_levels,
-      typename InitialGridPoints::type initial_number_of_grid_points) noexcept;
+      "given by tensor products of the specified BlockBounds. Each Block in\n"
+      "the lattice is identified by a VolumeDim-tuple of zero-based indices\n"
+      "Supplying a list of these tuples to BlocksToExclude will result in\n"
+      "the domain having the corresponding Blocks excluded. See the Domain\n"
+      "Creation tutorial in the documentation for more information on Block\n"
+      "numberings in rectilinear domains. Note that if any Blocks are\n"
+      "excluded, the option IsPeriodicIn is ignored as periodic boundary\n"
```
Comment (nilsdeppe) [inline src/Domain/Creators/AlignedLattice.hpp]: Could you error if is_periodic and `ignore_blocks` are set?

**Review by nilsdeppe — COMMENTED** [review]
One small comment, feel free to squash immediately :)

**Comment by nilsdeppe** [comment]
Marking priority because this is needed for further disk debugging.

File: src/Domain/Creators/AlignedLattice.cpp [hunk 2]
```diff
@@ -20,36 +21,58 @@ AlignedLattice<VolumeDim, TargetFrame>::AlignedLattice(
...
     // clang-tidy: trivially copyable
     : block_bounds_(std::move(block_bounds)),         // NOLINT
       is_periodic_in_(std::move(is_periodic_in)),     // NOLINT
       initial_refinement_levels_(                     // NOLINT
           std::move(initial_refinement_levels)),      // NOLINT
       initial_number_of_grid_points_(                 // NOLINT
           std::move(initial_number_of_grid_points)),  // NOLINT
-      number_of_blocks_by_dim_{map_array(block_bounds_, [
-      ](const std::vector<double>& v) noexcept { return v.size() - 1; })} {}
+      blocks_to_exclude_(blocks_to_exclude),          // NOLINT
```
Comment (nilsdeppe) [inline src/Domain/Creators/AlignedLattice.cpp]: 1. Don't NOLINT
2. `std::move`

I think as a general thing going forward we should avoid using things like `typename OPTION::type blah` in the constructors because it obscures the types. I think if the types in the constructor were `double`, `std::vector`, etc. it would be much easier to understand when to move and when not to. I think this should be a separate PR, though :)

File: src/Domain/Creators/AlignedLattice.cpp [hunk 3]
```diff
@@ -20,36 +21,58 @@ AlignedLattice<VolumeDim, TargetFrame>::AlignedLattice(
     const typename BlockBounds::type block_bounds,
     const typename IsPeriodicIn::type is_periodic_in,
     const typename InitialRefinement::type initial_refinement_levels,
-    const typename InitialGridPoints::type
-        initial_number_of_grid_points) noexcept
+    const typename InitialGridPoints::type initial_number_of_grid_points,
+    const typename BlocksToExclude::type blocks_to_exclude) noexcept
```
Comment (nilsdeppe) [inline src/Domain/Creators/AlignedLattice.cpp]: make non-const

File: src/Domain/Creators/AlignedLattice.cpp [hunk 4]
```diff
@@ -20,36 +21,58 @@ AlignedLattice<VolumeDim, TargetFrame>::AlignedLattice(
...
           std::move(initial_number_of_grid_points)),  // NOLINT
-      number_of_blocks_by_dim_{map_array(block_bounds_, [
-      ](const std::vector<double>& v) noexcept { return v.size() - 1; })} {}
+      blocks_to_exclude_(blocks_to_exclude),          // NOLINT
+      number_of_blocks_by_dim_{
+          map_array(block_bounds_, [](const std::vector<double>& v) noexcept {
+            return v.size() - 1;
+          })} {
+  if (not blocks_to_exclude_.empty()) {
+    for (const auto& dim_is_periodic : is_periodic_in_) {
```
Comment (nilsdeppe) [inline src/Domain/Creators/AlignedLattice.cpp]: I think using STL algorithms here is a good choice:

```cpp
if (not blocks_to_exclude_.empty() and 
    alg::any_of(is_periodic_in_, [](const bool t) noexcept { return t; })) {
  ERROR(...);
}
```

**Review by nilsdeppe — COMMENTED** [review]
Minor comments that'll make clang-tidy happy and a suggestion on using an STL algorithm to simplify some code. I'm happy for you to squash :)

File: tests/Unit/Domain/Creators/Test_AlignedLattice.cpp [hunk 5]
```diff
@@ -65,5 +65,49 @@ SPECTRE_TEST_CASE("Unit.Domain.Creators.AlignedLattice", "[Domain][Unit]") {
       dynamic_cast<const creators::AlignedLattice<3, Frame::Inertial>*>(
           domain_creator_3d.get());
   test_aligned_blocks(*aligned_blocks_creator_3d);
+
+  const auto rubiks_cube_with_hole_domain =
```
Comment (hrueter) [inline tests/Unit/Domain/Creators/Test_AlignedLattice.cpp]: Do you call this rubiks cube, because it is supposed to be 3x3x3?

Will we run into trademark issues, when we use this name?

File: src/Domain/Creators/AlignedLattice.cpp (hunk 3 above)
Comment (mar-celine) [reply src/Domain/Creators/AlignedLattice.cpp]: This is so that `std::move` actually does something right? Should the other parameters be made non const as well? This can be made an issue, but I'll fix the line I added.

File: tests/Unit/Domain/Creators/Test_AlignedLattice.cpp (hunk 5 above)
Comment (mar-celine) [reply tests/Unit/Domain/Creators/Test_AlignedLattice.cpp]: I'd hope not, but I'll make the change to `cubical_shell` regardless :+1:

**Review by nilsdeppe — APPROVED** [review]

**Review by kidder — APPROVED** [review]
